packages/plugin-core/src/commands/MoveHeader.ts

```diff
@@ -0,0 +1,311 @@
+      throw this.headerNotSelectedError;
+    }
+    const tree = proc.parse(maybeNote.body);
+    let headerFound = false;
+    let foundHeaderIndex: number | undefined;
+    let nextHeaderIndex: number | undefined;
+    visit(tree, (node, index) => {
+      if (nextHeaderIndex) {
+        return;
+      }
+      const depth = node.depth as Heading["depth"];
+      if (!headerFound) {
+        if (node.type === DendronASTTypes.HEADING) {
+          if (
+            depth === targetHeader!.depth &&
+            this.hasIdenticalChildren(node, targetHeader!)
+          ) {
+            headerFound = true;
+            foundHeaderIndex = index;
+            return;
+          }
+        }
+      } else if (node.type === DendronASTTypes.HEADING) {
+        if (foundHeaderIndex) {
+          if (depth <= targetHeader!.depth) nextHeaderIndex = index;
+        }
+      }
+    });
+
+    const nodesToMove = nextHeaderIndex
+      ? (tree.children as any[]).splice(
+          foundHeaderIndex!,
+          nextHeaderIndex! - foundHeaderIndex!
+        )
+      : (tree.children as any[]).splice(foundHeaderIndex!);
+
+    const lookupController = LookupControllerV3.create({
+      nodeType: "note",
+      disableVaultSelection: true,
+    });
+    const lookupProvider = new NoteLookupProvider(this.key, {
+      allowNewNote: false,
+      noHidePickerOnAccept: false,
+    });
+
+    lookupController.show({
+      title: "Select note to move header to",
+      placeholder: "note",
+      provider: lookupProvider,
+      initialValue: opts?.dest ? opts.dest.fname : undefined,
+      nonInteractive: !_.isUndefined(opts?.dest),
+    });
+
+    return new Promise((resolve) => {
+      HistoryService.instance().subscribev2("lookupProvider", {
+        id: this.key,
+        listener: async (event) => {
+          if (event.action === "done") {
+            HistoryService.instance().remove(this.key, "lookupProvider");
+            const cdata = event.data as NoteLookupProviderSuccessResp;
+            resolve({
+              dest: cdata.selectedItems[0],
+              origin: maybeNote,
+              nodesToMove,
+              modifiedOriginTree: tree,
+              originProc: proc,
+              engine,
+            });
+            lookupController.onHide();
+          } else if (event.action === "error") {
+            this.L.error({ msg: `error: ${event.data}` });
+            resolve(undefined);
+          } else if (event.action === "changeState") {
+            this.L.info({ msg: "cancelled" });
+            resolve(undefined);
+          } else {
+            this.L.error({ msg: `unhandled error: ${event.data}` });
+            resolve(undefined);
+          }
+        },
+      });
+    });
+  }
+
+  removePosition = (nodes: Node[]) => {
+    return _.map(nodes, (node) => {
+      visit(node, (n) => {
+        delete n["position"];
+      });
+      return node;
+    });
+  };
+
+  async execute(opts: CommandOpts): Promise<CommandOutput> {
```

Comment:
This isn't necessarily request to change this function, but more so kick off some discussions on the preferred code style, that we typically gravitate towards.

This in particular talking about the typical length of functions that we are ok with. Some rules of thumbs that we can consider:
* keeping the length of functions to allow entire function to be able to fit on a single screen. 
* Anytime there is a comment describing a block of code, ask whether it makes sense to split that code block into a function, name that function as the comment and get rid of the comment. 

For example as broadstrokes refactoring this function could have some splits added as:

```
 async execute(opts: CommandOpts): Promise<CommandOutput> {
    const ctx = "MoveHeaderCommand";
    this.L.info({ ctx, opts });
    const {
      dest,
      origin,
      nodesToMove,
      modifiedOriginTree,
      originProc,
      engine,
    } = opts;

    // deep copy the origin before mutating it
    const originDeepCopy = _.cloneDeep(origin);
    // remove header from origin
    const modifiedOriginContent = originProc.stringify(modifiedOriginTree);

    origin.body = modifiedOriginContent;
    await engine.writeNote(origin, {
      updateExisting: true,
    });

    // append header to destination
    await this.appendHeaderToDestination(engine, dest, nodesToMove);

    delayedUpdateDecorations();

    // update reference
    const anchorsToUpdateNames = this.findAnchorsToUpdate(originDeepCopy, modifiedOriginContent);

    const foundReferences = await findReferences(origin.fname);
    const updated: NoteProps[] = [];

    _.forEach(foundReferences, async (reference) => {
      const { location, isCandidate } = reference;
      // ignore candidate refs.
      if (isCandidate) return false;
      const note = this.getNoteByLocation(location, engine);
      if (!note) return false;

      // find match text in note and modify the node
      const proc = this.getProc(engine, note);
      const tree = proc.parse(note.body);
      visit(tree, (node: Node) => {
        if (node.type === DendronASTTypes.WIKI_LINK) {
          if (
            node.value === origin.fname &&
            _.includes(anchorsToUpdateNames, node.data!.anchorHeader)
          ) {
            node.value = dest!.fname;
          }
        }
      });
      note.body = proc.stringify(tree);
      const writeResp = await engine.writeNote(note, { updateExisting: true });
      updated.push(writeResp.data[0].note);
      return;
    });
    return { ...opts, updated };
  }

  private async appendHeaderToDestination(engine: EngineAPIService, dest: NoteProps | undefined, nodesToMove: Node[]) {
    const destProc = this.getProc(engine, dest!);
    const destTree = destProc.parse(dest!.body);
    destTree.children = (destTree.children as Node[]).concat(nodesToMove);
    const modifiedDestContent = destProc.stringify(destTree);
    dest!.body = modifiedDestContent;
    await engine.writeNote(dest!, {
      updateExisting: true,
    });
  }

  private findAnchorsToUpdate(originDeepCopy: NoteProps, modifiedOriginContent: string) {
    const anchorsBefore = RemarkUtils.findAnchors(originDeepCopy.body);
    const anchorsAfter = RemarkUtils.findAnchors(modifiedOriginContent);
    const anchorsToUpdate = _.differenceWith(
      anchorsBefore,
      anchorsAfter,
      this.hasIdenticalChildren
    );
    const anchorsToUpdateNames = _.map(anchorsToUpdate, (anchor) => {
      const slugger = getSlugger();
      const payload = AnchorUtils.anchorNode2anchor(anchor, slugger);
      return payload![0];
    });
    return anchorsToUpdateNames;
  }

  private getNoteByLocation(location: any, engine: EngineAPIService) {
    const fsPath = location.uri.fsPath;
    const {vaults, wsRoot, notes} = engine;
    const fname = NoteUtils.normalizeFname(path.basename(fsPath));
    const vault = VaultUtils.getVaultByNotePath({
      fsPath,
      wsRoot,
      vaults,
    });
    const note = NoteUtils.getNoteByFnameV5({
      fname,
      notes,
      vault,
      wsRoot,
    });
    return note;
  }
  ```

Reply:
That is a good point. After reviewing a couple of your PRs I sensed that you take a lot of effort in commenting your code which I found to be useful when reviewing. I was planning to add some coding style guide regarding this, but I have been deferring that for a bit 😅 

I definitely agree on the points you made here. A rule of thumb in terms of splitting code into separate functions right now that our team follows is to split it when it makes sense to generalize it so that it could be reused somewhere else. (but I've seen things get split up even when it wasn't reused somewhere else, for readability).